<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Numpy-Documentation" data-toc-modified-id="Numpy-Documentation-1">Numpy Documentation</a></span></li><li><span><a href="#Why-Numpy" data-toc-modified-id="Why-Numpy-2">Why Numpy</a></span><ul class="toc-item"><li><span><a href="#Speed" data-toc-modified-id="Speed-2.1">Speed</a></span></li><li><span><a href="#Vectorized-Operation" data-toc-modified-id="Vectorized-Operation-2.2">Vectorized Operation</a></span></li><li><span><a href="#Other-Benefits" data-toc-modified-id="Other-Benefits-2.3">Other Benefits</a></span></li></ul></li><li><span><a href="#Create-NumPy-ndarrays" data-toc-modified-id="Create-NumPy-ndarrays-3">Create NumPy ndarrays</a></span><ul class="toc-item"><li><span><a href="#Use-np.array()-to-Convert-an-Object,-e.g.-List" data-toc-modified-id="Use-np.array()-to-Convert-an-Object,-e.g.-List-3.1">Use <code>np.array()</code> to Convert an Object, e.g. List</a></span><ul class="toc-item"><li><span><a href="#1D-array" data-toc-modified-id="1D-array-3.1.1">1D array</a></span></li><li><span><a href="#2D-Array" data-toc-modified-id="2D-Array-3.1.2">2D Array</a></span></li><li><span><a href="#Upcasting-array.dtype" data-toc-modified-id="Upcasting-array.dtype-3.1.3">Upcasting <code>array.dtype</code></a></span></li><li><span><a href="#Save-to-&amp;-load-from-.npy-file" data-toc-modified-id="Save-to-&amp;-load-from-.npy-file-3.1.4">Save to &amp; load from <code>.npy</code> file</a></span></li></ul></li><li><span><a href="#Use-Numpy-Built-in-Functions-to-Create-ndarrays" data-toc-modified-id="Use-Numpy-Built-in-Functions-to-Create-ndarrays-3.2">Use Numpy Built-in Functions to Create ndarrays</a></span><ul class="toc-item"><li><span><a href="#Create-ndarrays-with-constants" data-toc-modified-id="Create-ndarrays-with-constants-3.2.1">Create ndarrays with constants</a></span></li><li><span><a href="#Create-ndarrays-with-special-main-diagonal-values" data-toc-modified-id="Create-ndarrays-with-special-main-diagonal-values-3.2.2">Create ndarrays with special main diagonal values</a></span></li><li><span><a href="#Create-ndarrays-with-a-range-of-evenly-spaced-ints" data-toc-modified-id="Create-ndarrays-with-a-range-of-evenly-spaced-ints-3.2.3">Create ndarrays with a range of evenly spaced ints</a></span></li><li><span><a href="#Create-ndarrays-with-a-range-of-evenly-spaced-floats" data-toc-modified-id="Create-ndarrays-with-a-range-of-evenly-spaced-floats-3.2.4">Create ndarrays with a range of evenly spaced floats</a></span></li><li><span><a href="#Reshape-ndarrays" data-toc-modified-id="Reshape-ndarrays-3.2.5">Reshape ndarrays</a></span></li><li><span><a href="#Create-ndarrays-of-random-values" data-toc-modified-id="Create-ndarrays-of-random-values-3.2.6">Create ndarrays of random values</a></span></li></ul></li><li><span><a href="#Quiz" data-toc-modified-id="Quiz-3.3">Quiz</a></span></li></ul></li><li><span><a href="#Accessing-&amp;-Modifying-ndarray-Elements" data-toc-modified-id="Accessing-&amp;-Modifying-ndarray-Elements-4">Accessing &amp; Modifying ndarray Elements</a></span><ul class="toc-item"><li><span><a href="#1D-Array" data-toc-modified-id="1D-Array-4.1">1D Array</a></span></li><li><span><a href="#2D-Array" data-toc-modified-id="2D-Array-4.2">2D Array</a></span></li></ul></li><li><span><a href="#Deleting,-Adding-or-Inserting-Values-to-ndarray" data-toc-modified-id="Deleting,-Adding-or-Inserting-Values-to-ndarray-5">Deleting, Adding or Inserting Values to ndarray</a></span><ul class="toc-item"><li><span><a href="#np.delete" data-toc-modified-id="np.delete-5.1"><code>np.delete</code></a></span><ul class="toc-item"><li><span><a href="#1D-Array" data-toc-modified-id="1D-Array-5.1.1">1D Array</a></span></li><li><span><a href="#2D-Array" data-toc-modified-id="2D-Array-5.1.2">2D Array</a></span></li></ul></li><li><span><a href="#np.append" data-toc-modified-id="np.append-5.2"><code>np.append</code></a></span><ul class="toc-item"><li><span><a href="#1D-Array" data-toc-modified-id="1D-Array-5.2.1">1D Array</a></span></li><li><span><a href="#2D-Array" data-toc-modified-id="2D-Array-5.2.2">2D Array</a></span></li></ul></li><li><span><a href="#np.insert" data-toc-modified-id="np.insert-5.3"><code>np.insert</code></a></span><ul class="toc-item"><li><span><a href="#1D-Array" data-toc-modified-id="1D-Array-5.3.1">1D Array</a></span></li><li><span><a href="#2D-Array" data-toc-modified-id="2D-Array-5.3.2">2D Array</a></span></li></ul></li></ul></li><li><span><a href="#Stacking-arrays" data-toc-modified-id="Stacking-arrays-6">Stacking arrays</a></span><ul class="toc-item"><li><span><a href="#vstack" data-toc-modified-id="vstack-6.1"><code>vstack</code></a></span><ul class="toc-item"><li><span><a href="#hstack" data-toc-modified-id="hstack-6.1.1"><code>hstack</code></a></span></li></ul></li></ul></li><li><span><a href="#Slicing-ndarrays" data-toc-modified-id="Slicing-ndarrays-7">Slicing ndarrays</a></span><ul class="toc-item"><li><span><a href="#1D-array" data-toc-modified-id="1D-array-7.1">1D array</a></span></li><li><span><a href="#2D-array" data-toc-modified-id="2D-array-7.2">2D array</a></span></li><li><span><a href="#Create-a-new-ndarray-using-slicing-with-np.copy" data-toc-modified-id="Create-a-new-ndarray-using-slicing-with-np.copy-7.3">Create a new ndarray using slicing with <code>np.copy</code></a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Use-np.copy-to-create-new-ndarray-with-slicing" data-toc-modified-id="Use-np.copy-to-create-new-ndarray-with-slicing-7.3.0.1">Use <code>np.copy</code> to create new ndarray with slicing</a></span></li></ul></li></ul></li><li><span><a href="#Slicing-with-special-built-in-functions" data-toc-modified-id="Slicing-with-special-built-in-functions-7.4">Slicing with special built-in functions</a></span><ul class="toc-item"><li><span><a href="#np.diag(ndarray,-k=N)" data-toc-modified-id="np.diag(ndarray,-k=N)-7.4.1"><code>np.diag(ndarray, k=N)</code></a></span></li><li><span><a href="#np.unique(ndarray)" data-toc-modified-id="np.unique(ndarray)-7.4.2"><code>np.unique(ndarray)</code></a></span></li></ul></li></ul></li><li><span><a href="#Boolean-Indexing" data-toc-modified-id="Boolean-Indexing-8">Boolean Indexing</a></span></li><li><span><a href="#Set-Operations" data-toc-modified-id="Set-Operations-9">Set Operations</a></span><ul class="toc-item"><li><span><a href="#np.intersect1d,-np.setdiff1d,-np.union1d" data-toc-modified-id="np.intersect1d,-np.setdiff1d,-np.union1d-9.1"><code>np.intersect1d</code>, <code>np.setdiff1d</code>, <code>np.union1d</code></a></span></li></ul></li><li><span><a href="#Sorting-ndarrays" data-toc-modified-id="Sorting-ndarrays-10">Sorting ndarrays</a></span><ul class="toc-item"><li><span><a href="#1D-array" data-toc-modified-id="1D-array-10.1">1D array</a></span><ul class="toc-item"><li><span><a href="#np.sort-vs-ndarray.sort" data-toc-modified-id="np.sort-vs-ndarray.sort-10.1.1"><code>np.sort</code> vs <code>ndarray.sort</code></a></span></li></ul></li><li><span><a href="#2D-array" data-toc-modified-id="2D-array-10.2">2D array</a></span></li></ul></li><li><span><a href="#Arithmetic-Operations" data-toc-modified-id="Arithmetic-Operations-11">Arithmetic Operations</a></span><ul class="toc-item"><li><span><a href="#Element-wise" data-toc-modified-id="Element-wise-11.1">Element-wise</a></span><ul class="toc-item"><li><span><a href="#1D-Array" data-toc-modified-id="1D-Array-11.1.1">1D Array</a></span></li><li><span><a href="#2D-array" data-toc-modified-id="2D-array-11.1.2">2D array</a></span></li></ul></li><li><span><a href="#Matrix-wise-(not-discussed)" data-toc-modified-id="Matrix-wise-(not-discussed)-11.2">Matrix-wise (not discussed)</a></span></li></ul></li><li><span><a href="#Broadcasting" data-toc-modified-id="Broadcasting-12">Broadcasting</a></span><ul class="toc-item"><li><span><a href="#1D-array" data-toc-modified-id="1D-array-12.1">1D array</a></span></li><li><span><a href="#2D-array" data-toc-modified-id="2D-array-12.2">2D array</a></span></li><li><span><a href="#Quiz" data-toc-modified-id="Quiz-12.3">Quiz</a></span></li></ul></li></ul></div>

In [2]:
import numpy as np

# Numpy Documentation

- [Numpy Manual - Scipy.org](https://docs.scipy.org/doc/numpy-1.13.0/contents.html)

- [Numpy User Guide - Scipy.org](https://docs.scipy.org/doc/numpy-1.13.0/user/index.html)

- [Numpy Reference - Scipy.org](https://docs.scipy.org/doc/numpy-1.13.0/reference/index.html#reference)

- [Numpy Lectures - Scipy.org](http://www.scipy-lectures.org/intro/numpy/index.html)

# Why Numpy

## Speed

In [11]:
# create an array of 100 million floats between 0 and 1
x=np.random.random(100000000)

In [10]:
%%time
# calculate mean using Python builtin functions
sum(x)/len(x)

CPU times: user 14.3 s, sys: 54.3 ms, total: 14.4 s
Wall time: 14.4 s


0.49998640408012135

In [14]:
%%time
# calculate mean using np.mean
np.mean(x)

CPU times: user 127 ms, sys: 2.15 ms, total: 129 ms
Wall time: 128 ms


0.5000861143570895

In [13]:
# How many times faster is `np.mean()`?
14.4/0.12

120.00000000000001

## Vectorized Operation

In [24]:
# create another array of 100 million floats
y=np.random.random(100000000)

In [26]:
# to get the product of x and y in Numpy is extreamly easy 
# with vectorized operation, instead of using nested loop in standard bPython
z = x*y
np.mean(z)

0.25003328085552906

## Other Benefits

- Numpy's **multidimensional array data structures** that can represent **vectors** and **matrices** : machine learning is based on lots of matrix multiplication


- Optimized **built-in functions** for **complex mathematical computations**: fast and very little code to write


- `Pandas` is built on top of `Numpy`.



# Create NumPy ndarrays

## Use `np.array()` to Convert an Object, e.g. List


### 1D array

In [53]:
x = np.array([1,2,3,4,5,6])

print('{}\n'.format(x))
print('dtype: {}'.format(x.dtype))
print('shape: {}'.format(x.shape))
print('size: {}'.format(x.size))

[1 2 3 4 5 6]

dtype: int64
shape: (6,)
size: 6


### 2D Array

In [52]:
y = np.array([[1,2,3],[4,5,6],[7,8,9],[10,11,12]])
print('{}\n'.format(y))
print('dtype: {}'.format(y.dtype))
print('shape: {}'.format(y.shape))
print('size: {}'.format(y.size))

[[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]]

dtype: int64
shape: (4, 3)
size: 12


### Upcasting `array.dtype`

In [36]:
# For an array with both float and int, Numpy upcast int to float to preserve
# preciseness
x = np.array([0.5, 1, 3, 4])
x.dtype

dtype('float64')

In [38]:
# To define an array with dtype
x = np.array([0.5, 1, 3, 4], dtype=int)
x.dtype

dtype('int64')

### Save to & load from `.npy` file 

In [43]:
# save an array to `.npy` file in the current directory
np.save('my_array', x)

In [42]:
# load a `.npy` file in the current directory to an array
x = np.load('my_array.npy')
x

array([0, 1, 3, 4])

## Use Numpy Built-in Functions to Create ndarrays


### Create ndarrays with constants

In [56]:
np.zeros((3,4), dtype=int)

array([[0, 0, 0, 0],
       [0, 0, 0, 0],
       [0, 0, 0, 0]])

In [57]:
np.ones((4,5))

array([[1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.]])

In [58]:
np.full((4,3), 5)

array([[5, 5, 5],
       [5, 5, 5],
       [5, 5, 5],
       [5, 5, 5]])

### Create ndarrays with special main diagonal values

In [59]:
# identity matrix - a square matrix in which all the elements 
# of the principal diagonal are ones and all other elements are zeros.
np.eye(5)

array([[1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1.]])

In [61]:
# use a sequence of value to be the main the main diagnol of
# a square matrix, and fill in the rest with zeros
np.diag([10, 20, 30, 50])

array([[10,  0,  0,  0],
       [ 0, 20,  0,  0],
       [ 0,  0, 30,  0],
       [ 0,  0,  0, 50]])

### Create ndarrays with a range of evenly spaced ints

In [62]:
# Create an array of 10 integers from 0 to 10
np.arange(10)

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [63]:
# Create an array from 4 to 10
np.arange(4,10)

array([4, 5, 6, 7, 8, 9])

In [65]:
# Create an array from 1 to 14 with step as 3
np.arange(1,14,3)

array([ 1,  4,  7, 10, 13])

### Create ndarrays with a range of evenly spaced floats

In [92]:
# for non-integer steps, use `np.linspace(start, stop, n)`
# n = 50 by default, if unspecified 
# Create an array of 50 evenly spaced floats from 0 to 25, 25 inclusive
np.linspace(0, 25)

array([ 0.        ,  0.51020408,  1.02040816,  1.53061224,  2.04081633,
        2.55102041,  3.06122449,  3.57142857,  4.08163265,  4.59183673,
        5.10204082,  5.6122449 ,  6.12244898,  6.63265306,  7.14285714,
        7.65306122,  8.16326531,  8.67346939,  9.18367347,  9.69387755,
       10.20408163, 10.71428571, 11.2244898 , 11.73469388, 12.24489796,
       12.75510204, 13.26530612, 13.7755102 , 14.28571429, 14.79591837,
       15.30612245, 15.81632653, 16.32653061, 16.83673469, 17.34693878,
       17.85714286, 18.36734694, 18.87755102, 19.3877551 , 19.89795918,
       20.40816327, 20.91836735, 21.42857143, 21.93877551, 22.44897959,
       22.95918367, 23.46938776, 23.97959184, 24.48979592, 25.        ])

In [70]:
# Create 10 evenly spaced floats from 0 to 25, 25 non-inclusive
np.linspace(0, 25, 10, endpoint=False)

array([ 0. ,  2.5,  5. ,  7.5, 10. , 12.5, 15. , 17.5, 20. , 22.5])

### Reshape ndarrays

In [93]:
# `np.reshape` as function
x = np.linspace(0, 25, 10)
np.reshape(x, (5,2))

array([[ 0.        ,  2.77777778],
       [ 5.55555556,  8.33333333],
       [11.11111111, 13.88888889],
       [16.66666667, 19.44444444],
       [22.22222222, 25.        ]])

In [94]:
# `.reshape` as method of an array
x.reshape(2,5)

array([[ 0.        ,  2.77777778,  5.55555556,  8.33333333, 11.11111111],
       [13.88888889, 16.66666667, 19.44444444, 22.22222222, 25.        ]])

In [77]:
np.arange(20).reshape((4,5))

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14],
       [15, 16, 17, 18, 19]])

### Create ndarrays of random values

In [95]:
# Create random floats from 0 to 1, use tuple to define the shape of ndarrays
np.random.random((3,3))

array([[0.93853525, 0.79340273, 0.17561358],
       [0.70394778, 0.81710677, 0.20048461],
       [0.35356292, 0.74888748, 0.42180301]])

In [96]:
# Create random integers with low-bound, upbound and shape
np.random.randint(4,15,(3,2))

array([[11, 11],
       [13,  9],
       [ 5,  5]])

In [89]:
# Create random normal distribution
# create a 1000 x 1000 normal distribution with mean = 0, std = 0.1
np.random.seed(999)
x = np.random.normal(0,0.1, size=(1000, 1000))
print('{}\n'.format(x))
print('mean: {}'.format(x.mean()))
print('std: {}'.format(x.std()))
print('max: {}'.format(x.max()))
print('min: {}'.format(x.min()))
print('# positive: ', (x>0).sum())
print('# negative: ', (x<0).sum())

[[ 0.01271578  0.14018909  0.0314815  ...  0.08816739 -0.16914039
  -0.05118324]
 [ 0.14681943  0.10643196 -0.07934169 ... -0.02493764 -0.04698768
  -0.10235213]
 [-0.13220536 -0.06862131  0.01862244 ... -0.05284694  0.14744174
   0.02426232]
 ...
 [ 0.10052159 -0.03756336 -0.03854788 ... -0.10713039 -0.07383442
  -0.03971498]
 [-0.05022672 -0.06837407  0.03034411 ...  0.0987338  -0.01282658
   0.05142451]
 [-0.07253547 -0.06069139  0.04499498 ... -0.11027383 -0.06382385
  -0.11737963]]

mean: -1.8185526754438272e-05
std: 0.09988707339870906
max: 0.49346807945662463
min: -0.47306382149916043
# positive:  500213
# negative:  499787


The mean is $ -1.81855 \dot 10^{-5}$ which is $0.0000181855$, estimated as `0`

## Quiz

Using the Built-in functions you learned, create a 4 x 4 ndarray that only contains consecutive even numbers from 2 to 32 (inclusive)

In [97]:
X = np.arange(2,33,2).reshape((4,4))
X

array([[ 2,  4,  6,  8],
       [10, 12, 14, 16],
       [18, 20, 22, 24],
       [26, 28, 30, 32]])

# Accessing & Modifying ndarray Elements

## 1D Array

In [107]:
# Access 1D array elements like list
x = np.arange(1,10)
print(x)
print('first element:', x[0])
print('second element:', x[1])
print('last element:', x[-1])
print('second last element:', x[-2])
x[0]=3
print(x)

[1 2 3 4 5 6 7 8 9]
first element: 1
second element: 2
last element: 9
second last element: 8
[3 2 3 4 5 6 7 8 9]


## 2D Array

In [108]:
x = np.arange(1,10).reshape((3,3))
print(x)
print('Element at [0,0]:', x[0,0])
print('Element at [0,1]:', x[0,1])
print('Element at [2,2]:', x[2,2])
x[0,0]=3
print(x)

[[1 2 3]
 [4 5 6]
 [7 8 9]]
Element at [0,0]: 1
Element at [0,1]: 2
Element at [2,2]: 9
[[3 2 3]
 [4 5 6]
 [7 8 9]]


# Deleting, Adding or Inserting Values to ndarray


## `np.delete`

`np.delete(ndarray, elements, axis)`

### 1D Array

In [143]:
x = np.arange(1,10).reshape((3,3))
print("Original array: \n", x, '\n')

# delete the first and last element of x
x_1 = np.delete(x, [0, 4])
print("First and last element deleted:\n", x_1, "\n")

Original array: 
 [[1 2 3]
 [4 5 6]
 [7 8 9]] 

First and last element deleted:
 [2 3 4 6 7 8 9] 



### 2D Array

In [144]:
# delete first row of x
x_2 = np.delete(x, 0, axis=0)
print("First row deleted: \n", x_2, "\n")


# delete the first and last column of x
x_3 = np.delete(x, [0,2], axis=1)
print("First and last column deleted: \n", x_3)

First row deleted: 
 [[4 5 6]
 [7 8 9]] 

First and last column deleted: 
 [[2]
 [5]
 [8]]


## `np.append`

`np.append(ndarray, elements, axis)`

In [146]:
x = np.arange(1,6)
y = np.arange(1,7).reshape((2,3))

### 1D Array

In [145]:
# append a value
print("x:\n", x, '\n')
print("\nmodified x:\n", np.append(x, 6))

x:
 [[1 2 3]
 [4 5 6]
 [7 8 9]] 


modified x:
 [1 2 3 4 5 6 7 8 9 6]


In [142]:
# append multiple values
print("x:\n", x, '\n')
print("\nmodified x:\n",np.append(x, [6,7]))

x:
 [1 2 3 4 5] 


modified x:
 [1 2 3 4 5 6 7]


### 2D Array

In [137]:
# append a new row
print("y:\n", y)
print("\nmodified y: \n", np.append(y, [[7,8,9]], axis=0))

y:
 [[1 2 3]
 [4 5 6]]

modified y: 
 [[1 2 3]
 [4 5 6]
 [7 8 9]]


In [161]:
# append a new column
print("y:\n", y)
print("\nmodified y: \n", np.append(y, [[9],[10]], axis=1))

y:
 [[1 2 3]
 [4 5 6]]

modified y: 
 [[ 1  2  3  9]
 [ 4  5  6 10]]


## `np.insert`

### 1D Array

In [149]:
x = np.arange(1,6)

# insert 8 and 9 at the third element
print("x:\n", x, '\n')
print("\nmodified x:\n",np.insert(x, 2, [8,9]))

x:
 [1 2 3 4 5] 


modified x:
 [1 2 8 9 3 4 5]


### 2D Array

In [152]:
y = np.arange(1,7).reshape((2,3))

# insert [7,8,9] into the second row of the matrix (2D array)
print("y:\n", y)
print("\nmodified y: \n", np.insert(y, 1, [[7,8,9]], axis=0))

y:
 [[1 2 3]
 [4 5 6]]

modified y: 
 [[1 2 3]
 [7 8 9]
 [4 5 6]]


In [156]:
# insert [9, 10] into the second column of the matrix
print("y:\n", y)
print("\nmodified y: \n", np.insert(y, 1, [9,10], axis=1))

y:
 [[1 2 3]
 [4 5 6]]

modified y: 
 [[ 1  9  2  3]
 [ 4 10  5  6]]


# Stacking arrays

## `vstack`

In [163]:
x = np.array([1,2])
y = np.array([[3,4], [5,6]])

In [167]:
# vstack x and y
print("x: \n", x)
print("\ny: \n", y)
print("\nstacked vertically:\n", np.vstack((x,y)))

x: 
 [1 2]

y: 
 [[3 4]
 [5 6]]

stacked vertically:
 [[1 2]
 [3 4]
 [5 6]]


### `hstack`

In [171]:
# hstack x and y
# in order to stack horizontally, we need to reshape x as 2 by 1 first
print("x: \n", x)
print("\ny: \n", y)
print("\nstacked horizontally:\n", np.hstack((y, x.reshape(2,1))))

x: 
 [1 2]

y: 
 [[3 4]
 [5 6]]

stacked horizontally:
 [[3 4 1]
 [5 6 2]]


# Slicing ndarrays

## 1D array

```python
1. ndarray[start:end]
2. ndarray[start:]
3. ndarray[:end]
```

## 2D array


```python
ndarray[row_start:row_end, col_start:col_end]
```

In [181]:
x = np.arange(1,21).reshape(4,5)
print(x,'\n')

[[ 1  2  3  4  5]
 [ 6  7  8  9 10]
 [11 12 13 14 15]
 [16 17 18 19 20]] 



In [183]:
# Select elements in 2nd - 4th rows and 3rd-5th cols
print(x[1:4, 2:5])

[[ 8  9 10]
 [13 14 15]
 [18 19 20]]


In [184]:
# Select elements in 2nd - last row and 3rd - last cols
print(x[1:,2:])

[[ 8  9 10]
 [13 14 15]
 [18 19 20]]


In [178]:
# Select elements in first 3 rows and 3rd - last cols
print(x[:3, 2:])

[[ 3  4  5]
 [ 8  9 10]
 [13 14 15]]


In [180]:
# Select all elements of 3rd row
print(x[2, :])

[11 12 13 14 15]


In [179]:
# Select all elements of 3rd col - result as 1D array
print(x[:, 2])

[ 3  8 13 18]


In [185]:
# Select all elements of 3rd col - result as 2D array
print(x[:, 2:3])

[[ 3]
 [ 8]
 [13]
 [18]]


## Create a new ndarray using slicing with `np.copy` 

Assigning a sliced ndarray to a new variable will actually just create a new variable points to the same object as the old variable. **Modifying the new variable will also modifies the old variable too**.

In [186]:
z = x[1:4, 2:5]
print('x:\n', x, '\n')
print('z:\n', z)

x:
 [[ 1  2  3  4  5]
 [ 6  7  8  9 10]
 [11 12 13 14 15]
 [16 17 18 19 20]] 

z:
 [[ 8  9 10]
 [13 14 15]
 [18 19 20]]


In [187]:
# Modifying z will also modifies x
z[2,2]=555
print('x:\n', x, '\n')
print('z:\n', z)

x:
 [[  1   2   3   4   5]
 [  6   7   8   9  10]
 [ 11  12  13  14  15]
 [ 16  17  18  19 555]] 

z:
 [[  8   9  10]
 [ 13  14  15]
 [ 18  19 555]]


#### Use `np.copy` to create new ndarray with slicing

This will create a new variable pointing to a new object created from the old sliced ndarray.

In [188]:
z = np.copy(x[1:4, 2:5])
print('x:\n', x, '\n')
print('z:\n', z)

x:
 [[  1   2   3   4   5]
 [  6   7   8   9  10]
 [ 11  12  13  14  15]
 [ 16  17  18  19 555]] 

z:
 [[  8   9  10]
 [ 13  14  15]
 [ 18  19 555]]


In [189]:
z[2,2]=50
print('x:\n', x, '\n')
print('z:\n', z)

x:
 [[  1   2   3   4   5]
 [  6   7   8   9  10]
 [ 11  12  13  14  15]
 [ 16  17  18  19 555]] 

z:
 [[ 8  9 10]
 [13 14 15]
 [18 19 50]]


## Slicing with special built-in functions

### `np.diag(ndarray, k=N)` 

The function extracts the elements along the diagonal defined by `N`. 
- As default is `k=0`, which refers to the main diagonal. 
- Values of `k > 0` are used to select elements in diagonals above the main diagonal, and 
- values of `k < 0` are used to select elements in diagonals below the main diagonal.

In [196]:
# Select the elements in the main diagonal
x = np.arange(25).reshape(5,5)
print('x:\n', x, '\n')
print('on main diagonal:\n', np.diag(x))

x:
 [[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]
 [15 16 17 18 19]
 [20 21 22 23 24]] 

on main diagonal:
 [ 0  6 12 18 24]


In [195]:
# Select the elements above the main diagonal
print('x:\n', x, '\n')
print('above main diagonal:\n', np.diag(x, k=1))

x:
 [[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]
 [15 16 17 18 19]
 [20 21 22 23 24]] 

above main diagonal:
 [ 1  7 13 19]


In [194]:
# Select the elements below the main diagonal
print('x:\n', x, '\n')
print('below main diagonal:\n', np.diag(x, k=-1))

x:
 [[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]
 [15 16 17 18 19]
 [20 21 22 23 24]] 

below main diagonal:
 [ 5 11 17 23]


### `np.unique(ndarray)`

The function extracts only the unique elements in an ndarray.

In [197]:
x = np.array([[1,2,3],[5,2,8],[1,2,3]])
print('x:\n', x, '\n')
print('unique elements in x:\n', np.unique(x))

x:
 [[1 2 3]
 [5 2 8]
 [1 2 3]] 

unique elements in x:
 [1 2 3 5 8]


# Boolean Indexing

In [214]:
x = np.arange(25).reshape(5, 5)

print()
print('x:\n', x, '\n')

# to select elements in X:
print('greater than 10:\n', x[x > 10])
print('less than or equal to 7:\n', x[x <= 7])
print('between 10 and 17:\n', x[(x > 10) & (x < 17)])


x:
 [[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]
 [15 16 17 18 19]
 [20 21 22 23 24]] 

greater than 10:
 [11 12 13 14 15 16 17 18 19 20 21 22 23 24]
less than or equal to 7:
 [0 1 2 3 4 5 6 7]
between 10 and 17:
 [11 12 13 14 15 16]


In [215]:
# to assign the elements between 10 and 17 as -1
x[(x > 10) & (x < 17)] = -1
print(x)

[[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 -1 -1 -1 -1]
 [-1 -1 17 18 19]
 [20 21 22 23 24]]


# Set Operations

## `np.intersect1d`, `np.setdiff1d`, `np.union1d`

This is useful when **comparing ndarrays**, for example, to find common elements between two ndarrays. 

In [220]:
x = np.array([1,2,3,4,5])
y = np.array([6,7,2,8,4])


print('x:\n{}\n'.format(x))
print('y:\n{}\n'.format(y))

# set operations to compare x and y
print('both in x and y:', np.intersect1d(x,y))
print('in x that are not in y:', np.setdiff1d(x,y))
print('x and y:',np.union1d(x,y))

x:
[1 2 3 4 5]

y:
[6 7 2 8 4]

both in x and y: [2 4]
in x that are not in y: [1 3 5]
x and y: [1 2 3 4 5 6 7 8]


# Sorting ndarrays



## 1D array

### `np.sort` vs `ndarray.sort` 

`np.sort(ndarray)` - it won't modify the order of the original ndarray.

In [236]:
x = np.random.randint(1,11, size=(10,))
print('x: \n{}\n'.format(x))
print('sorted:\n{}\n'.format(np.sort(x)))
print('x after sorting: \n{}\n'.format(x))

x: 
[6 8 6 3 7 4 7 6 5 1]

sorted:
[1 3 4 5 6 6 6 7 7 8]

x after sorting: 
[6 8 6 3 7 4 7 6 5 1]



`ndarray.sort()` - it will modify the order of the ndarray.

In [238]:
x = np.random.randint(1,11, size=(10,))
print('x: \n{}\n'.format(x))

x.sort()
print('x after sorting: \n{}\n'.format(x))

x: 
[3 1 7 1 4 1 5 2 3 3]

x after sorting: 
[1 1 1 2 3 3 3 4 5 7]



## 2D array

In [248]:
x = np.random.randint(1,11, size=(10,1))
print('x: \n{}\n'.format(x))

# sorted rows
print('sorted:\n{}\n'.format(np.sort(x, axis=0)))

x: 
[[ 6]
 [10]
 [ 5]
 [ 5]
 [ 7]
 [ 1]
 [ 7]
 [ 5]
 [ 7]
 [ 5]]

sorted:
[[ 1]
 [ 5]
 [ 5]
 [ 5]
 [ 5]
 [ 6]
 [ 7]
 [ 7]
 [ 7]
 [10]]



In [246]:
x = np.random.randint(1,11, size=(5,2))
print('x: \n{}\n'.format(x))

# sorted rows
print('sorted:\n{}\n'.format(np.sort(x, axis=0)))

x: 
[[ 4 10]
 [ 9  4]
 [ 8  8]
 [ 7  1]
 [ 9  1]]

sorted:
[[ 4  1]
 [ 7  1]
 [ 8  4]
 [ 9  8]
 [ 9 10]]



In [247]:
x = np.random.randint(1,11, size=(5,5))
print('x: \n{}\n'.format(x))

# sorted columns
print('sorted:\n{}\n'.format(np.sort(x, axis=1)))

x: 
[[ 9  6  5  5  4]
 [ 9  8  1  9  5]
 [ 4  7  1  7  6]
 [ 7  8  4 10  8]
 [ 6  7  1  2  2]]

sorted:
[[ 4  5  5  6  9]
 [ 1  5  8  9  9]
 [ 1  4  6  7  7]
 [ 4  7  8  8 10]
 [ 1  2  2  6  7]]



# Arithmetic Operations

## Element-wise 

### 1D Array

In [259]:
x = np.array([1,2,3,4])
y = np.array([5.5,6.5,7.5,8.5])
print('x:\n{}\n'.format(x))
print('y:\n{}\n'.format(y))

# element-wise operation using arithmetic symbols and function
print('x + y = ', x + y)
print('add(x,y) = ', np.add(x,y))
print()
print('x - y = ', x - y)
print('subtract(x,y) = ', np.subtract(x,y))
print()
print('x * y = ', x * y)
print('multiply(x,y) = ', np.multiply(x,y))
print()
print('x / y = ', x / y)
print('divide(x,y) = ', np.divide(x,y))
print()

# apply math function to all elements in 1D array
print('EXP(x) =', np.exp(x))
print()
print('SQRT(x) =',np.sqrt(x))
print()
print('POW(x,2) =',np.power(x,2)) # We raise all elements to the power of 2

x:
[1 2 3 4]

y:
[5.5 6.5 7.5 8.5]

x + y =  [ 6.5  8.5 10.5 12.5]
add(x,y) =  [ 6.5  8.5 10.5 12.5]

x - y =  [-4.5 -4.5 -4.5 -4.5]
subtract(x,y) =  [-4.5 -4.5 -4.5 -4.5]

x * y =  [ 5.5 13.  22.5 34. ]
multiply(x,y) =  [ 5.5 13.  22.5 34. ]

x / y =  [0.18181818 0.30769231 0.4        0.47058824]
divide(x,y) =  [0.18181818 0.30769231 0.4        0.47058824]

EXP(x) = [ 2.71828183  7.3890561  20.08553692 54.59815003]

SQRT(x) = [1.         1.41421356 1.73205081 2.        ]

POW(x,2) = [ 1  4  9 16]


### 2D array

In [265]:
X = np.array([1,2,3,4]).reshape(2,2)
Y = np.array([5.5,6.5,7.5,8.5]).reshape(2,2)
print('X:\n{}\n'.format(X))
print('Y:\n{}\n'.format(Y))

# element-wise operations using arithmetic symbols and functions
print('X + Y = \n', X + Y)
print()
print('add(X,Y) = \n', np.add(X,Y))
print()
print('X - Y = \n', X - Y)
print()
print('subtract(X,Y) = \n', np.subtract(X,Y))
print()
print('X * Y = \n', X * Y)
print()
print('multiply(X,Y) = \n', np.multiply(X,Y))
print()
print('X / Y = \n', X / Y)
print()
print('divide(X,Y) = \n', np.divide(X,Y))


# apply math functions in 2D array:
print("\n======= Apply math functions in 2D array:===========\n")
print('EXP(x):\n{}\n'.format(np.exp(X)))
print('SQRT(x):\n{}\n'.format(np.sqrt(X)))
print('POW(x,2):\n{}\n'.format(np.power(X,2)))


# apply stat functions in 2D array:
print("\n======= Apply stat functions in 2D array:===========\n")
print('Average of all elements in X:', X.mean())
print('Average of all elements in the columns of X:', X.mean(axis=0))
print('Average of all elements in the rows of X:', X.mean(axis=1))
print()
print('Sum of all elements in X:', X.sum())
print('Sum of all elements in the columns of X:', X.sum(axis=0))
print('Sum of all elements in the rows of X:', X.sum(axis=1))
print()

X:
[[1 2]
 [3 4]]

Y:
[[5.5 6.5]
 [7.5 8.5]]

X + Y = 
 [[ 6.5  8.5]
 [10.5 12.5]]

add(X,Y) = 
 [[ 6.5  8.5]
 [10.5 12.5]]

X - Y = 
 [[-4.5 -4.5]
 [-4.5 -4.5]]

subtract(X,Y) = 
 [[-4.5 -4.5]
 [-4.5 -4.5]]

X * Y = 
 [[ 5.5 13. ]
 [22.5 34. ]]

multiply(X,Y) = 
 [[ 5.5 13. ]
 [22.5 34. ]]

X / Y = 
 [[0.18181818 0.30769231]
 [0.4        0.47058824]]

divide(X,Y) = 
 [[0.18181818 0.30769231]
 [0.4        0.47058824]]

======= Apply math functions in 2D array:===========

EXP(x):
[[ 2.71828183  7.3890561 ]
 [20.08553692 54.59815003]]

SQRT(x):
[[1.         1.41421356]
 [1.73205081 2.        ]]

POW(x,2):
[[ 1  4]
 [ 9 16]]


======= Apply stat functions in 2D array:===========

Average of all elements in X: 2.5
Average of all elements in the columns of X: [2. 3.]
Average of all elements in the rows of X: [1.5 3.5]

Sum of all elements in X: 10
Sum of all elements in the columns of X: [4 6]
Sum of all elements in the rows of X: [3 7]



## Matrix-wise (not discussed)


# Broadcasting

## 1D array

In [279]:
X = np.arange(1,5).reshape(2,2)
print()
print('X = \n', X)
print()

print('3 * X = \n', 3 * X)
print()
print('3 + X = \n', 3 + X)
print()
print('X - 3 = \n', X - 3)
print()
print('X / 3 = \n', X / 3)


X = 
 [[1 2]
 [3 4]]

3 * X = 
 [[ 3  6]
 [ 9 12]]

3 + X = 
 [[4 5]
 [6 7]]

X - 3 = 
 [[-2 -1]
 [ 0  1]]

X / 3 = 
 [[0.33333333 0.66666667]
 [1.         1.33333333]]


NumPy is working behind the scenes to **broadcast `3` along the ndarray so that they have the same shape**. This allows us to add `3` to each element of `X` with just one line of code.

## 2D array
Subject to certain constraints, Numpy can do the same for two ndarrays of different shapes, as we can see below.

In [284]:
# x is a 1 x 3 row (MUST BE the same row size as y)
x = np.arange(1,4)

# y is a 4 x 3 array
y = np.arange(1,13).reshape(4,3)

# Z is a 4 x 1 column (MUST BE the same column size as y)
z = np.arange(1,5).reshape(4,1)
print('x:\n{}\n'.format(x))
print('y:\n{}\n'.format(y))
print('z:\n{}\n'.format(z))

# Numpy broadcast x or y along y to have the same shape (3 x 3)
print('x + y = \n', x + y)
print()
print('z + y = \n',z + y)

x:
[1 2 3]

y:
[[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]]

z:
[[1]
 [2]
 [3]
 [4]]

x + y = 
 [[ 2  4  6]
 [ 5  7  9]
 [ 8 10 12]
 [11 13 15]]

z + y = 
 [[ 2  3  4]
 [ 6  7  8]
 [10 11 12]
 [14 15 16]]


## Quiz

Use Broadcasting to create a 4 x 4 ndarray that has its first column full of 1s, its second column full of 2s, its third column full of 3s, etc.. 

In [286]:
X = np.ones((4,4))*np.arange(1,5)
print(X)

[[1. 2. 3. 4.]
 [1. 2. 3. 4.]
 [1. 2. 3. 4.]
 [1. 2. 3. 4.]]
